### ISSUE TITLE
serial-getty for the kernel boot console shouldn't have additional baud rates other than the one set in cmdline

### ISSUE THREAD

**Issue by jjakob** [systemd/systemd#15611]
**Is your feature request related to a problem? Please describe.**
Currently systemd-getty-generator auto-spawns a serial-getty@.service if the cmdline's 'console=' parameter has a serial console set. The spawned agetty has options `--keep-baud 115200,38400,9600`, so the initial baud rate will be the one set in the cmdline, but can be changed to the other baud rates by sending a BREAK character over the serial console.

The problem is that when plugging and unplugging serial cables, break characters are often received due to the noise injected into RXD when the data lines connect before ground. This is very often (in my case it was almost always) when using a TRS jack connection, but is possible with standard D-sub serial cables too if the shell isn't grounded.

This means that the simple act of connecting or disconnecting the serial cable can make the console switch baud rates, making the console unresponsive. If the user knows about the issue, he can try sending breaks to cycle back to the expected baud rate, but that's surely not the expected behavior.

**Describe the solution you'd like**
When the cmdline's 'console=' has a serial console and baud rate set, the auto-generated serial-getty@ttySx.service should start agetty just with `--keep-baud` and no list of alternative baud rates.
Maybe serial-getty@.service should be modified so it behaves that way in all cases, not just for the main system console.

**Describe alternatives you've considered**
I currently use a override, /etc/systemd/system/serial-getty@ttyS0.service.d/override.conf:
```
[Service]
ExecStart=
ExecStart=-/sbin/agetty -o '-p -- \\u' 57600 %I $TERM
```
In my case, I always want a fixed baud rate of 57600, but that can be simply exchanged with `--keep-baud` to work for everyone without modifying.
This is just to get the behavior I would expect should be by default.

**Comment by poettering** [comment]
Well, this sounds like something of a dilemma. Some people might say that keep-baud is the only robust thing, and others might say that iterating through the baud values is much more robust. I am not sure we can find a globally perfect solution for this. One will be better for some, the other for others, I doubt we can make everybody happy here...

**Comment by kjetilho** [comment]
The main problem with BREAK cycling through baud rates is that it eats the BREAK character, and this makes it impossible to send magic sysrq commands via ILO.  We disable it for this reason.

I don't think it is a worthwhile use case to be able to cycle to a working getty when kernel command line serial configuration is wrong.  The first step to fix that computer should be to adjust boot parameters in GRUB (temporarily) so that you can read boot output *and* get a working getty.

A possible compromise would be to enable this cycling only when booting in rescue mode.

**Comment by jjakob** [comment]
Well, I actually can send a sysrq even with baud-cycling (multiple baud rates passed to agetty), using screen: press `C-a C-b <sysrq char>` and agetty receives the sysrq fine without cycling. I actually had trouble getting it to cycle when I wanted it to: `C-a C-b` didn't send a break no matter how many times I repeated it, so I had to do `C-a : break <enter>` TWICE to get it to switch. Which is in contrast to plugging the cable, where it switches almost always.

I think there is a bigger issue with agetty itself: once cycled from the initial (kernel, --keep-baud) baud rate, **it doesn't cycle back to it.** It just cycles through the bauds passed to it via parameters. This is definitely a problem, in addition to the main system console having additional baud rates set via the service file.

**Comment by karelzak** [comment]
I think @jjakob is right. It seems good idea to remember the original speed and use it also when cycle in the list of available speeds. I have changed agetty to do it by default on --keep-baud.

**Comment by poettering** [comment]
hmm, can we close this here now that agetty cycles through the `--keep-baud` rate too?

**Comment by jjakob** [comment]
That does not change this issue which is that the main boot console should, in my opinion, only have the one single speed set (--keep-baud or fixed at the one from cmdline's console)

**Comment by poettering** [comment]
Well, again, I am pretty sure changing that would just make things less robust for others even though it makes it more robust for you. I see no clear indication that one approach is clearly better than the other...

**Comment by jjakob** [comment]
We must have different definitions of "robust" then - in my definition, robust means that if I set console=ttyS0,57600 in cmdline, I expect the console to be that speed, always. if just plugging in the cable causes it to switch to a different baud rate, I need to send breaks until I get back to the baud rate I set it to and expected it to stay at.
This is different than "convenience" or "user-friendliness" - though it's debatable if requiring the user to always cycle through baud rates to get a usable console is really "friendly" to them.
This cycling is also very different than "auto-baud" detection, which is IMO much more user-friendly than cycling, but agetty doesn't support it AFAIK.

**Comment by michaelolbrich** [comment]
@jjakob the problem is, that relying on the kernel to set the baud rate only works for serial consoles. If you start a serial-getty on anything else then you get some arbitrary baud rate, probably 9600. And in this case, being able to switch to a different baud rate is very useful.

@poettering maybe we need a different template for the serial-getty that is automatically started for the kernel console. Because for this specific use-case I agree that always keeping the baud rate specified in the kernel commandline is correct.

**Comment by kjetilho** [comment]
yes, to be clear - I only want this change in behaviour on the kernel console, not on other serial consoles.

**Comment by poettering** [comment]
@michaelolbrich yupp have to distinct units sounds OK to me. happy to review/merge a PR that add serial-getty-console@.service and makes the getty generator use it.

**Comment by laf0rge** [comment]
I strongly argue that the serial boot console should have one fixed baudarate only, and that it should not eat BREAK or any other characters.  This has been the standard behavior that I assume from working with linux serial consoles from embedded to data center servers for ~25 years.
